crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,696 @@
+use crate::{
+    authority::{
+        Authority, LookupControlFlow, LookupError, LookupObject, LookupOptions, MessageRequest,
+        UpdateResult, ZoneType,
+    },
+    proto::{
+        op::{Query, ResponseCode},
+        rr::{
+            rdata::{A, AAAA, TXT},
+            LowerName, Name, RData, Record, RecordType,
+        },
+    },
+    resolver::lookup::Lookup,
+    server::RequestInfo,
+    store::blocklist::{BlocklistConfig, BlocklistConsultAction},
+};
+
+// TODO:
+//  * Add (optional) support for logging the client IP address.  This will require some Authority
+//    trait changes to accomplish.
+//  * Add query-type specific results for non-address queries.
+//  * Add support for per-blocklist sinkhole IPs, block messages, actions.
+//  * Add support for an exclusion list: allow the user to configure a list of patterns that
+//    will never be insert into the in-memory blocklist (such as their own domain.)
+//  * Add support for regex matching
+
+/// A conditional authority that will resolve queries against one or more block lists and return a
+/// forged response.  The typical use case will be to use this in a chained configuration before a
+/// forwarding or recursive resolver in order to pre-emptively block queries for hosts that are on
+/// a block list. Refer to tests/test-data/test_configs/chained_blocklist.toml for an example
+/// of this configuration.
+///
+/// The blocklist authority also supports the consult interface, which allows an authority to review
+/// a query/response that has been processed by another authority, and, optionally, overwrite that
+/// response before returning it to the requestor.  There is an example of this configuration in
+/// tests/test-data/test_configs/example_consulting_blocklist.toml.  The main intended use of this
+/// feature is to allow log-only configurations, to allow administrators to see if blocklist domains
+/// are being queried.  While this can be configured to overwrite responses, it is not recommended
+/// to do so - it is both more efficient, and more secure, to allow the blocklist to drop queries
+/// pre-emptively, as in the first example.
+pub struct BlocklistAuthority {
+    origin: LowerName,
+    blocklist: HashMap<LowerName, bool>,
+    wildcard_match: bool,
+    min_wildcard_depth: u8,
+    sinkhole_ipv4: Ipv4Addr,
+    sinkhole_ipv6: Ipv6Addr,
+    ttl: u32,
+    block_message: Option<String>,
+    consult_action: BlocklistConsultAction,
+}
+
+impl BlocklistAuthority {
+    /// Read the Authority for the origin from the specified configuration
+    pub async fn try_from_config(
+        origin: Name,
+        _zone_type: ZoneType,
+        config: &BlocklistConfig,
+        base_dir: Option<&Path>,
+    ) -> Result<Self, String> {
+        info!("loading blocklist config: {origin}");
+
+        let sinkhole_ipv4 = if let Some(ip) = config.sinkhole_ipv4 {
+            ip
+        } else {
+            Ipv4Addr::from_str("0.0.0.0").unwrap()
```

Comment:
Suggest spelling this as:

```rust
let sinkhole_ipv4 = match config.sinkhole_ipv4 {
    Some(ip) => ip,
    None => Ipv4Addr::UNSPECIFIED,
};
```

(And similar for IPv6.)

I'd also consider just inlining these into the `Self` initializiation below.